add extension point for placing symbols to custom places.

Anipik opened the pull request:
## Bug

No good way to place symbols outside of the lib folder

Fixes: https://github.com/NuGet/Home/issues/10913

Related to : https://github.com/NuGet/Home/issues/10860

Currently there was no way to add extra symbols files in the innerbuilds, this change provide that hook. It still wont address the problem of targetpath being an absolute path. I added the tests which should help in understanding the workflow for innerbuilds.


## Description

- added a variable  TargetsForTfmSpecificDebugSymbolsInPackage  which could be set to run the targets to populate _TargetPathsToSymbolsWithTfm 
-  Moved IncludeBuildOutput added here https://github.com/NuGet/NuGet.Client/pull/3810 in itemgroup in order to allow the user to put the buildOutput\symbols at some other place other than lib. the same pattern is used from other targets as well like _AddPriFileToPackBuildOutput and _GetFrameworksWithSuppressedDependencies

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
- [x] needs to be added


- **Documentation**
  - [x] Documentation issue https://github.com/NuGet/Home/issues/10891

cc @ViktorHofer @safern @ericstj @nkolev92

Anipik committed:
add extension point for symbol

ericstj commented:
@Anipik can you explain what the user can do to accomplish the following with this PR?

1. From the outer build include a file at runtimes/rid/native/foo.pdb?  What item and metadata should I set?  Will the PDB be absent from the main package and present in the symbols package?
2. From inner build, define that my primary output should be placed at dotnet/analyzers/cs/myAnalyzer.dll and dotnet/analyzers/cs/myAnalyzer.pdb.

Thanks!

nkolev92 reviewed (commented):
The change seems reasonable. 

Can you give an example as Eric suggested. 

Please make sure you add a test.

ericstj commented:
It's possible this fix doesn't address everything in my suggestion, but if that's the case be sure not to close that issue.  It's still important to have a clear statement of what this fixes and how to use it.

erdembayar commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -433,9 +433,8 @@ Copyright (c) .NET Foundation. All rights reserved.
-          Condition="'$(IncludeBuildOutput)' == 'true'"
           Returns="@(BuildOutputInPackage)">
-    <ItemGroup>
+    <ItemGroup Condition="'$(IncludeBuildOutput)' == 'true'">
```
@Anipik
Thank you for your contribution. Could you be able to add unit tests? 
Most likely below 2 are best places put your new tests: https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskTests.cs
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs

Anipik replied:
Yes i am working on adding tests. i should be able to push the tests change today.

Anipik committed:
add tests

Anipik commented:
@ericstj @nkolev92 i added the tests and updated the description as well. let me know if u still have any questions.

@erdembayar @nkolev92 feel free to trigger the ci.

erdembayar commented:
> @ericstj @nkolev92 i added the tests and updated the description as well. let me know if u still have any questions.
> 
> @erdembayar @nkolev92 feel free to trigger the ci.

@Anipik 
Ok. Approved for CI build. 🤞

ericstj commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                    var xml = XDocument.Load(stream);
+                    var target = @"<Target Name=""CustomContentTarget"">
+    <ItemGroup>
+      <_TargetPathsToSymbolsWithTfm Include=""abc.pdb"">
```
Are we OK with having this be a private item that we expect people to set?  Should you instead append to `DebugSymbolsProjectOutputGroupOutput` or do we worry that has side-effects?

Anipik replied:
Yeah the main concern here was to introduce the side effects if we rename the property. cc @nkolev92

ericstj replied:
I wasn't suggesting any rename, just that we suggest that folks use the existing item already used by NuGet: https://github.com/NuGet/NuGet.Client/blob/25e7298a33a3ce861457cf590c76105b2e4e9cb2/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L464

ericstj replied:
I guess that wouldn't work well if they also tried to set `IncludeBuildOutput` to `false`.  It seems unusual to me that we would create a public property `TargetsForTfmSpecificDebugSymbolsInPackage` that requires the use of a private item `_TargetPathsToSymbolsWithTfm`.  Should we create `TfmSpecificDebugSymbolsFile` item instead?  cc @nkolev92

nkolev92 replied:
I think create a new item is a good idea.

Quickly skimming, there might not be any consequences in renaming that item. When calling it, NuGet uses https://github.com/NuGet/NuGet.Client/blob/775c3bc68ff105930920fc1f8fd6abd8b88d54db/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L380-L382.

ericstj replied:
Renaming will break people who were using the private item, why not just include a public item into `_TargetPathsToSymbolsWithTfm`.  Non breaking and behaves like https://github.com/NuGet/NuGet.Client/blob/25e7298a33a3ce861457cf590c76105b2e4e9cb2/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L452

nkolev92 replied:
Sure, we can do that too.

ViktorHofer replied:
GitHub shows 85 consumers of the internal item: https://github.com/search?q=_TargetPathsToSymbolsWithTfm&type=code. That's probably reason enough to not rename it.

nkolev92 replied:
> GitHub shows 85 consumers of the internal item: https://github.com/search?q=_TargetPathsToSymbolsWithTfm&type=code. That's probably reason enough to not rename it.

If you look further, a large portion of those are `NuGet.Build.Tasks.Pack.targets` :D

ViktorHofer replied:
Right there is some noise in the query but there are still a lot of valid hits. Multiply that with an imaginary factor to include customer code that isn't on GitHub and you will likely prefer to not rename :)

ericstj commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                    var target = @"<Target Name=""CustomContentTarget"">
+    <ItemGroup>
+      <_TargetPathsToSymbolsWithTfm Include=""abc.pdb"">
+        <TargetPath>/runtimes/win/lib/$(TargetFramework)/abc.pdb</TargetPath>
```
Here's the usage that requires folks to "know" to use `TargetPath` and specify a rooted path.  It's the best we can do for now, but it's something we were calling out in https://github.com/NuGet/Home/issues/10860.  For now it's good that the test ensures that this stays working.

nkolev92 replied:
Can we create a follow up to `fix` this?

ericstj replied:
Yeah that issue remains open with this PR.  I think that one has a larger scope that what's being fixed here and I believe with the feedback this PR stands alone.

nkolev92 replied:
@Anipik Can you please follow up here.

Anipik replied:
yes as @ericstj mentioned we will need an independent fix for https://github.com/NuGet/Home/issues/10860 which will help us to not depend on targetPath. I will try to take a look later if i could find a good fix for this one as well.
However, both the issues are independent and I dont think there is any feedback here that needs to be addressed in this pr.

nkolev92 replied:
@Anipik Oh my bad, I didn't mean to suggest fix it. We should just file an issue to track the problem.

Anipik replied:
https://github.com/NuGet/NuGet.Client/pull/4053 issues mentions the problem of using the targetPath workaround.  and suggests to implement something similar to the TfmSpecificPackageFile

nkolev92 replied:
See https://github.com/NuGet/NuGet.Client/pull/4053#issuecomment-854057943. 

Resolving this thread.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                msbuildFixture.RestoreProject(workingDirectory, projectName, string.Empty);
```
No need for this. Restore is automatic. This just adds 1-2s to the test.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+    </ItemGroup>
+  </Target>";
+                    ProjectFileUtils.SetTargetFrameworkForProject(xml, tfmProperty, tfmValue);
+                    ProjectFileUtils.AddProperty(xml, "TargetsForTfmSpecificDebugSymbolsInPackage", "CustomContentTarget");
```
Can you create a follow up docs issue for documentating TargetsForTfmSpecificDebugSymbolsInPackage

erdembayar replied:
If you created doc issue then please paste link here.

Anipik replied:
do u want me to add the doc issue link to the codebase. the link is in the description

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                msbuildFixture.RestoreProject(workingDirectory, projectName, string.Empty);
```
no need to restore.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,105 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                    var files = nupkgReader.GetFiles();
+                    Assert.Contains(@"runtimes/win/lib/netstandard1.4/abc.pdb", files);
+                    Assert.Contains(@"runtimes/win/lib/net46/abc.pdb", files);
+                    Assert.DoesNotContain(@"lib/netstandard1.4/abc.pdb", files);
```
Is the idea that net46 should have a lib folder?
Should we validate that?

Anipik replied:
The idea here is that we can have custom build\symbol output without the actual build output. eg netstandard1.4 build output is not present but the custom symbol output (runtime file) is.

nkolev92 replied:
My comment was more about line 2797:

<!--StartFragment-->                    ProjectFileUtils.AddProperty(xml, "IncludeBuildOutput", "false", $"'$(TargetFramework)'=='netstandard1.4'");<br class="Apple-interchange-newline"><!--EndFragment-->


The conditional suggests to me that we expect something different for the different frameworks.

Anipik replied:
it seemed ```git commit ``` missed some of the files. Basically the build output files will be present for net46 and absent for netstandard1.4

nkolev92 reviewed (commented):
Just some minor feedback. On the right track!

The PR template originally had a field for a docs issue. Can you please add that back and create an issue?

Anipik committed:
remove restore and add net46 build output checks

Anipik committed:
missed change

Anipik commented:
@nkolev92 @erdembayar can you trigger the ci here ?

erdembayar commented:
> @nkolev92 @erdembayar can you trigger the ci here ?

Ok. Done.

Anipik commented:
Thanks :)

Anipik committed:
correct project name

Anipik committed:
add public property for users to set

ericstj commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -456,14 +455,16 @@ Copyright (c) .NET Foundation. All rights reserved.
       </_TargetPathsToSymbolsWithTfm>
     </ItemGroup>
+    <ItemGroup>
+      <_TargetPathsToSymbolsWithTfm Include="@(TfmSpecificDebugSymbolsFile)" />
```
Should this include the `NuGetRecursiveDir` metadata?  I couldn't quite follow from the source it that same path would be used for symbols or not: https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs#L646

Anipik replied:
cc @nkolev92

nkolev92 replied:
I don't think that codepath would be used, but would be cool to check out whether scenarios like the one described in  https://github.com/NuGet/NuGet.Client/pull/2117 work.

Anipik replied:
i tested it works. i can add a test for that as well.

nkolev92 replied:
Yes, please!

Anipik replied:
i pushed the test, can you trigger the ci ?

erdembayar replied:
Can you fix indent here? it looks whole xml tag off by1 space, of course it'll require another CI build.

Anipik replied:
The indentation seems right 
<img width="940" alt="Screen Shot 2021-06-01 at 11 03 32 AM" src="https://user-images.githubusercontent.com/10516582/120370240-2fbc0b80-c2c9-11eb-8299-da778e1dfca5.png">

erdembayar replied:
Which test actually covers https://github.com/NuGet/NuGet.Client/pull/2117 work?

Anipik replied:
PackCommand_ContentInnerTargetExtension_SymbolFilesDllWithRecursive covers some part of it. 
The automatic target path calculation wont work here because of this NuGet/Home#10860, hence we will to always specify the target path.

ericstj reviewed (approved):
LGTM modulo one comment.

Anipik committed:
add recursive directory test

Anipik commented:
@nkolev92 @erdembayar can we review and merge this one ?

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2718,6 +2718,155 @@ public void PackCommand_ContentInnerTargetExtension_AddsTfmSpecificContent(strin
+                using (var nupkgReader = new PackageArchiveReader(nupkgPath))
+                {
+                    var files = nupkgReader.GetFiles();
+                    Assert.Contains(@"runtimes/win/lib/netstandard1.4/abc.pdb", files);
```
Note that if you use snupkg, packages like this *won't* be accepted by nuget.org. 

Can you add some tests for snupkgs too?

Anipik replied:
added them. can you trigger the ci

Anipik committed:
add snupkg tests

nkolev92 reviewed (approved):
LGTM
Thanks for bearing with us adding all the tests!
One small comment left.

Great work!

nkolev92 commented:
@ericstj pointed out to me that https://github.com/NuGet/Home/issues/10860 contains *a lot* of detail, and might've been intended to be used as a follow up. 

We try to keep a 1-1 relationship between PR + issue. I created an issue that *this* PR will now close: https://github.com/NuGet/Home/issues/10913

ericstj commented:
Can this be merged now?

Anipik commented:
@nkolev92 @erdembayar can we merge this one :)
